flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,324 @@
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.io.network.partition.consumer.BufferProviderInputChannelBuilder;
+import org.apache.flink.runtime.io.network.partition.consumer.InputChannelID;
+import org.apache.flink.runtime.io.network.partition.consumer.RemoteInputChannel;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.embedded.EmbeddedChannel;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertTrue;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.BufferResponse;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.ErrorResponse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests the client side message decoder.
+ */
+public class NettyMessageClientDecoderDelegateTest {
+	private static final NettyBufferPool ALLOCATOR = new NettyBufferPool(1);
+
+	private static final InputChannelID NORMAL_CHANNEL_ID = new InputChannelID();
+
+	private static final InputChannelID RELEASED_CHANNEL_ID = new InputChannelID();
+
+	private static final InputChannelID REMOVED_CHANNEL_ID = new InputChannelID();
+
+	/**
+	 * Verifies that the client side decoder works well for unreleased input channels.
+	 */
+	@Test
+	public void testDownstreamMessageDecode() throws Exception {
+		// 3 buffers required.
+		testRepartitionMessagesAndDecode(3, false, false, false);
+	}
+
+	/**
+	 * Verifies that the client side decoder works well for empty buffers. Empty buffers should not
+	 * consume data buffers of the input channels.
+	 */
+	@Test
+	public void testDownstreamMessageDecodeWithEmptyBuffers() throws Exception {
+		// 4 buffers required.
+		testRepartitionMessagesAndDecode(4, true, false, false);
+	}
+
+	/**
+	 * Verifies that NettyMessageDecoder works well with buffers sent to a released  and removed input channels.
+	 * For such channels, no Buffer is available to receive the data buffer in the message, and the data buffer
+	 * part should be discarded before reading the next message.
+	 */
+	@Test
+	public void testDownstreamMessageDecodeWithReleasedAndRemovedInputChannel() throws Exception {
+		// 3 buffers required.
+		testRepartitionMessagesAndDecode(3, false, true, true);
+	}
+
+	//------------------------------------------------------------------------------------------------------------------
+
+	private void testRepartitionMessagesAndDecode(
+		int numberOfBuffersInNormalChannel,
+		boolean hasEmptyBuffer,
+		boolean hasBufferForReleasedChannel,
+		boolean hasBufferForRemovedChannel) throws Exception {
+
+		EmbeddedChannel channel = createPartitionRequestClientHandler(numberOfBuffersInNormalChannel);
+
+		try {
```

Comment:
`try (channel = createPartitionRequestClientHandler(numberOfBuffersInNormalChannel))`, then we can remove finally part.

Reply:
`EmbeddedChannel` seems not implemented `AutoClosable`.